### ISSUE TITLE
any moveto event prevents units to attack (GNA #22988)

### ISSUE THREAD

**Issue by wesnoth-bugs** [wesnoth/wesnoth#1427]
**[Original submission](https://gna.org/bugs/index.php?22988)** by slowthinker on 2014-11-23

In BfW a player can order his unit to attack a target unit that is not adjacent. 

(in a detail:

1) he clicks on his unit

2) he moves a mouse cursor over a hex that is adjacent to the target

3) he clicks on the target )

But this is not possible if the scenario contains a moveto event. In this case the attacking unit stops besides the target and must be re-ordered.

How to recreate: play this scenario:

```

[multiplayer]

	id=00 test no_direct_attack 

	name=_"00 test no_direct_attack"

	description=_ "testing no_direct_attack"

	map_data="{multiplayer/maps/2p_Den_of_Onis.map}"



	[event]

		name=moveto

		first_time_only=no

		[label]

			x,y=$unit.x,$unit.y

			text="moved"

		[/label]

		

	[/event] 	

[/multiplayer]

```

(Reproduced on Windows XP)
Release: 1.12.0
Priority: 5 - Normal
Severity: 3 - Normal

**Comment by wesnoth-bugs** [comment]
**Modified on 2014-11-24**

**gfgtdf wrote:**

> from looking at the code this seems to be the intended behaviour. In most situations this makes sense because the movto event might have changes the sitution, so that attacking is now a less good idea, and the user might want the option to rethink it.
> 
> 
> 
> an event with [allow_undo] should not cause this behaviour.

**Comment by wesnoth-bugs** [comment]
**Modified on 2014-11-24**

**slowthinker wrote:**

> Firstly, if this is not a bug then it is a new behaviour. I think UMC creators should be warned (maybe here? http://forums.wesnoth.org/viewtopic.php?f#21&t36057)
> 
> 
> 
> >In most situations this makes sense because the movto event might have changes the sitution
> 
> 
> 
> It might but it needn't to.
> 
> In my add-on I have a moveto event handler that does some work in the background only (e.g. recalculates income, because it depends on position of units), and so from the point of view of a player there is no such moveto event (i. e. a player doesn't know about it). Therefore a player must be surprised that suddenly the UI behaves differently.
> 
> 
> 
> >an event with [allow_undo] should not cause this behaviour
> 
> 
> 
> Indeed, I just tested it.
> 
> It sounds nice ... yes, an UMC author should have under his control whether the attack will be removed from the 'queue of orders' move+attack or not.
> 
> But what if the UMC author wants
> 
> [not to remove the attack] AND simultaneously [not to allow_undo]
> 
> ?
> 
> 
> 
> You can answer that it won't happen, because ...  
> 
> situations where no new information was revealed to the player in the moveto event handler (=A)
> 
> happen if and only if
> 
> the undo may be allowed (=B)
> 
> and it happens if and only if
> 
> there is no reason why the player would want to interrupt his order move+attack (in other words not to attack (=C)
> 
> 
> 
> so mathematically we can say A <#> B <> C
> 
> This idea sounds very reasonably.
> 
> (Also I noticed BfW 1.12 reflects this idea, because in 1.12 there is an "internal moveto event handler" that checks whether an enemy unit was revealed during the move and in that case the attack is canceled. (In BfW 1.10 the attack was performed anyway.) )
> 
> 
> 
> But ... I think "A <=> B" is not fully true: there may be a situation where "A and not B".
> 
> Example: a unit consumes 'fuel' during movement, and the fuel is stored in a variable. Then undo must be forbidden, because BfW offers no mechanism that would allow to restore that variable after undo (at least I don't know about that mechanism)

**Comment by wesnoth-bugs** [comment]
**Modified on 2014-11-24**

**slowthinker wrote:**

> slightly off topic: 
> 
> A mechanism that would support undo for UMC wouldn't be too complicated: two new events would be needed:
> 
> a) an event name=undo_waypoint_created that would be fired just after BfW created a new (way)point where the game may be undone to
> 
> b) an event name=undo_done that would be fired just after undo
> 
> (the waypoints would be numbered)
> 
> 
> 
> So the meaning of undo_waypoint_created would be:
> 
> wesnoth engine says to UMC: "A new undo_waypoint nr. XY has been created. You should backup your variables now, or you should start to register your next actions so that you may undo them when needed"
> 
> And the meaning of undo_done would be:
> 
> wesnoth engine says to UMC: "An undo has been done, and the game returned to the waypoint nr. YZ. You should restore your variables accordingly or you should undo the actions you registered"

**Comment by wesnoth-bugs** [comment]
**Modified on 2014-11-24**

**gfgtdf wrote:**

> no there is currently no mechanism for cutomizing undos, although it has been requested multiple times (see older feature requests, or this page: http://wiki.wesnoth.org/EasyCoding (i woudn't call this easy though)). 
> 
> 
> 
> currently [allow_undo] (or not allow undo) is the only thing that event "returns" to the c++ caller, i agree that this has quite some drawbacks in a situation like this. This is also annyong in enter/exit_hex events where [allow_undo] basicly means 'don't stop this unit in its move' and those events never prevent undoing. I think we should add a second tag [dont_interrupt] (do you know a better name?) that would then also solve the problems with enter/exit_hex.
> 
> 
> 
> However, this is not the part of the code im working on currently.

**Comment by wesnoth-bugs** [comment]
**Modified on 2014-11-24**

**slowthinker wrote:**

> >This is also annyong in enter/exit_hex events where [allow_undo] basicly means 'don't stop this unit in its move' and those events never prevent undoing.
> 
> you need not to allow_undo, because your UMC is unable to restore things
> 
> you need to allow_undo, because you want enter/exit_hex handlers don't stop the movement, and similarly moveto handlers don't cancel the next attack
> 
> 
> 
> so enter/exit hex is unusable in this situation, and players will cry they must split move+attack into two orders
> 
> 
> 
> >I think we should add a second tag [dont_interrupt] (do you know a better name?)
> 
> I was always missing a way how to cancel move/exit_hex/attack/hit/recruit... in the moment it is about to start.
> 
> If such a functionality was added then it could accomplish [interrupt] too.
> 
> 
> 
> ---------------------------------------
> 
> Again I am a bit off-topic, but this is another inconsistency of Wesnoth:
> 
> there are actions that change the game state, invoked by a player (move, attack, open a context menu, end turn ...); there are actions invoked by the Wesnoth engine (hit, stop movement, level unit ...)
> 
> And WML/lua is informed only about some actions and can control only some of them (e.g. a part of the context menu)

**Comment by wesnoth-bugs** [comment]
**Modified on 2014-11-24**

**gfgtdf wrote:**

> > so enter/exit hex is unusable in this situation, and players will cry they must split move+attack into two orders 
> 
> 
> 
> I don't know which situation you mean. If you are talking about the problm in the first post: I never said those events can help you in that situation.
> 
> Idk if you understand my intention correctly, but what i wanted to say is that if we would have a [dont_interrupt] wml action, we can change the meaning of [allow_undo] in those events to what it should mean: allowing undo and by default disallowing undos after enter/exit_hex events just like any other event.
> 
> 
> 
> > I was always missing a way how to cancel move/exit_hex/attack/hit/recruit... in the moment it is about to start. 
> 
> 
> 
> not for any user actions the wml event fires in a state we you can easily just interrupt the action and the 'abort' code would be needed to be impleemented for each event sperately, so most likeley that won't happen due to the work.
> 
> 
> 
> > And WML/lua is informed only about some actions and can control only some of them (e.g. a part of the context menu)
> 
> 
> 
> so you want to be informed about all context menu clicks (add label, enable planning mode)?

**Comment by wesnoth-bugs** [comment]
**Modified on 2014-11-25**

**slowthinker wrote:**

> >> so enter/exit hex is unusable in this situation, and players will cry they must split move+attack into two orders 
> 
> >I don't know which situation you mean.
> 
> By "situation" I mean
> 
> _"you need not to allow_undo, because your UMC is unable to restore things"_
> 
> 
> 
> So I think we talk about the same problem: [allow_undo] has two effects, and those should be split.
> 
> And yes, [dont_interrupt] would be a very simple solution. 
> 
> 
> 
> About the tag name: 
> 
> [dont_interrupt_order_queue] is more explanatory but it seems too long, and a good explanation in the wiki would be satisfatory probably
> 
> I have one objection though: "interrupt" sounds like it will continue later, maybe something like "cancel" is better? [dont_cancel] or [dont_cancel_orders] ...
> 
> 
> 
> (I will add more comments later)

**Comment by wesnoth-bugs** [comment]
**Modified on 2015-01-22**

**gfgtdf wrote:**

> Do you know whether 1.10 cancels the action if a moveto events appeares?
> 
> 
> 
> Another idea would be to make moveto events by default not cancel the current action and have a tag [cancel_action] instead.

**Comment by wesnoth-bugs** [comment]
**Modified on 2015-01-24**

**slowthinker wrote:**

> In BfW 1.10 (also 1.8. and 1.6) a moveto event cancels nothing. BfW 1.12 (1.11?) is a turning point. (sorry that I didn't stress it in the original post)
> 
> 
> 
> So yes, from a point of view of compatibility the default behaviour should not change and a tag [cancel_action] seems better.
> 
> 
> 
> a note: this behaviour might affect not only moveto, but more events, like exit_hex ... {?)

**Comment by gfgtdf** [comment]
fixed in 821e27c34f222c72d1cb02100370e0f7df2a75c2
